Enable native Windows 10 dark theme support by default

FranciscoPombal opened the pull request:
Settings related to workarounds to HiDPI issues
are still commented out by default.

---

See https://github.com/qbittorrent/qBittorrent/issues/13675#issuecomment-742888636 and Ctrl+F `darkmode` here: https://doc.qt.io/qt-5/qguiapplication.html

I did not test on a Windows version lower than Windows 10 1903, but I think in such cases Qt simply ignores this. However, it would be nice for someone to confirm this before merging, since we still support those versions.

The new font related argument that is commented out comes from here: https://github.com/qbittorrent/qBittorrent/issues/13749#issuecomment-742873508 and this whole discussion: https://github.com/qbittorrent/qBittorrent/issues/12295

Screenshots:

<details>
<summary>Click to expand</summary>

![dark mode ss1](https://user-images.githubusercontent.com/42386382/101896533-0edbee80-3ba1-11eb-813f-2898350c5ebb.png)
![dark mode ss5](https://user-images.githubusercontent.com/42386382/101896539-113e4880-3ba1-11eb-9ab0-65fb038a69e8.png)
![dark mode ss4](https://user-images.githubusercontent.com/42386382/101896544-126f7580-3ba1-11eb-9d8b-d208fede81c9.png)
![dark mode ss3](https://user-images.githubusercontent.com/42386382/101896547-13080c00-3ba1-11eb-83be-1cda11d28892.png)
![dark mode ss2](https://user-images.githubusercontent.com/42386382/101896549-13080c00-3ba1-11eb-99c2-a325705ee2b4.png)

It respects title bars and window borders color changes:

Go to Settings->Personalisation->Colours->"Choose your accent colour"
Show the accent colour on the following surfaces->Check/Enable Title bars and windows borders

![dark mode ss7](https://user-images.githubusercontent.com/42386382/101898803-a858cf80-3ba4-11eb-8ad1-e3b087c9a3f7.png)
![dark mode ss6](https://user-images.githubusercontent.com/42386382/101898808-aabb2980-3ba4-11eb-8bbf-7e54fd61c625.png)
</details>

xavier2k6 commented:
>darkmode=[1|2] controls how Qt responds to the activation of the Dark Mode for applications introduced in **Windows 10 1903** (since Qt 5.15).

>A value of 1 causes Qt to switch the window borders to black when Dark Mode for applications is activated and no High Contrast Theme is in use. This is intended for applications that implement their own theming.

>A value of 2 will in addition cause the Windows Vista style to be deactivated and switch to the Windows style using a simplified palette in dark mode. **This is currently experimental pending the introduction of new style that properly adapts to dark mode.**

https://doc.qt.io/qt-5/qguiapplication.html#platform-specific-arguments

glassez commented:
How does it interact with custom themes when both custom theme and Windows dark mode are enabled?

glassez commented:
Also please provide screenshots in PR description.

Chocobo1 commented:
I don't like the `fontengine` introduced, it is a temporary workaround at best. Its usage should be documented in wiki instead.

an0n666 commented:
Doesn't work for me.
20H2.

jagannatharjun commented on dist/windows/qt.conf:
```diff
@@ -2,4 +2,4 @@
 
 [Platforms]
-;WindowsArguments = dpiawareness=1
+WindowsArguments = darkmode=2;,dpiawareness=1,fontengine=freetype
```
may be use `darkmode=1;` 

> A value of 1 causes Qt to switch the window borders to black when Dark Mode for applications is activated and no High Contrast Theme is in use. This is intended for applications that implement their own theming.

we implement our own theming system

FranciscoPombal replied:
@jagannatharjun

I think what they mean in this case is applications that implement their own "forced" themeing (think Winamp/random early 2000s Windows applications). Our themeing is optional, so when no theme is in use, the conditions for `darkmode=2` apply (you can try `darkmode=1`, it does nothing).

Note: I have not yet tested what happens when both a custom theme is in use and `darkmode=2` is in effect. I would expect that the custom theme takes 100% precedence; after all, dark mode on Windows simply changes the base color palette, so it shouldn't matter what color the base color palette is when something goes on top of it.

xavier2k6 commented:
>Doesn't work for me.
20H2.

Go to Settings->Personalisation->Colours->"Dark"

**or**

Go to Settings->Personalisation->Colours->Custom->Choose your default app mode->Check/Enable "Dark"

xavier2k6 commented:
>controls how Qt responds to the **activation** of the **Dark Mode for applications**

Title is probably a bit misleading.....it's not a dark theme as such....

>Also please provide screenshots in PR description.

![qbitorrent dark mode](https://user-images.githubusercontent.com/42386382/101887001-337d9980-3b94-11eb-9f88-c0e7c590eff3.png)

jagannatharjun commented:
> 
![image](https://user-images.githubusercontent.com/34717789/101894549-562ba480-3bcc-11eb-8af2-3b9486c9de19.png)

This is clearly not ready for production

xavier2k6 commented:
some more screenshots:

![dark mode ss1](https://user-images.githubusercontent.com/42386382/101896533-0edbee80-3ba1-11eb-813f-2898350c5ebb.png)
![dark mode ss5](https://user-images.githubusercontent.com/42386382/101896539-113e4880-3ba1-11eb-9ab0-65fb038a69e8.png)
![dark mode ss4](https://user-images.githubusercontent.com/42386382/101896544-126f7580-3ba1-11eb-9d8b-d208fede81c9.png)
![dark mode ss3](https://user-images.githubusercontent.com/42386382/101896547-13080c00-3ba1-11eb-83be-1cda11d28892.png)
![dark mode ss2](https://user-images.githubusercontent.com/42386382/101896549-13080c00-3ba1-11eb-99c2-a325705ee2b4.png)

FranciscoPombal commented:
@jagannatharjun 

> This is clearly not ready for production

As the Qt docs state, it is currently experimental, so it can use some polish, but it's usable at least. I'm sure that even in this state, some users will prefer it. I'm expecting it to improve in future Qt versions of course. The user can always disable it for qBittorrent even if they have dark theme system-wide.

@xavier2k6 

> > controls how Qt responds to the **activation** of the **Dark Mode for applications**
> 
> Title is probably a bit misleading.....

How so? Enabling Qt to respond to the activation/deactivation of "dark mode for applications" is the same as saying that we enable _support_ for the native dark theme.

@Chocobo1 

> Adding `fontengine` should have its own commit.

Fixed. Let me know if you have a better suggestion for the commit message.

xavier2k6 commented:
>How so? Enabling Qt to respond to the activation/deactivation of "dark mode for applications" is the same as saying that we enable support for the native dark theme.

Apologies! that should've been part of this https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-743084256 (I think. there was a presumption that dark mode would've been on/active by default) not that there is a requirement of user interaction/choice....(my interpretation of the comment may be wrong too)

jagannatharjun commented:
> The user can always disable it for qBittorrent even if they have dark theme system-wide.

How? going and editing the .conf file is not good UX. IMHO darkmode=1; is more than enough for now.

FranciscoPombal commented:
@xavier2k6  @jagannatharjun 

> Apologies! that should've been part of this #13952 (comment) (I think. there was a presumption that dark mode would've been on/active by default) not that there is a requirement of user interaction/choice....(my interpretation of the comment may be wrong too)

Dark mode won't be on by default. Support for it will be active by default. This means that if the user selects "Dark mode for applications" in the Windows settings, qBittorrent will use dark mode. As soon as the user turns off "Dark mode for applications", qBittorrent automagically reverts back to the default native look & feel, without even needing to be restarted.

> How? going and editing the .conf file is not good UX. IMHO darkmode=1; is more than enough for now.

I doubt that a user who turns on "Dark mode for applications" in the Windows settings, which is meant to apply systemwide, prefers the non-dark native look to the dark theme, even if it does not have the greatest polish yet. Also, if they are truly unhappy with Qt's dark theme, I'm guessing they'd still much rather hunt for a custom dark theme first before resorting to deactivating the dark theme and use the non-dark native look. But hey, if they really want qBittorrent to be non-dark mode in their dark-mode system, the escape hatch is there in any case.

Also, again, as I've said before, `darkmode=1` is basically useless, it only affects the window borders, which is almost unnoticeable.

jagannatharjun commented:
>I doubt that a user who turns on "Dark mode for applications" in the Windows settings, which is meant to apply systemwide, prefers the non-dark native look to the dark theme, 

you know how bad it looks, right? and I assume you know we have our own theme system right? the whole point of our theming system is to overcome qt's limitations. why you want to eat half-baked cake when the chef himself advised not to eat.

> Also, again, as I've said before, darkmode=1 is basically useless, it only affects the window borders, which is almost unnoticeable

no with the custom theme this is more than enough https://github.com/jagannatharjun/qbt-theme/issues/40 I was thinking to open a PR over the weekend, huh

xavier2k6 commented:
>Dark mode won't be on by default. Support for it will be active by default.

I am aware of that. (we are getting our wires crossed, lol!) as explained - it should've been in response to an0n666's comment.

xavier2k6 commented:
>you know how bad it looks, right?

I think it's not too bad except for a few hiccups/colour changes required ("stalled arrows")

What has to be taken in to account as well is there will be users who won't want to use a 3rd party theme....

FranciscoPombal commented:
@glassez 

> How does it interact with custom themes when both custom theme and Windows dark mode are enabled?

As suspected in https://github.com/qbittorrent/qBittorrent/pull/13952#discussion_r540891122, it "just works" as expected: custom themes completely override the base dark theme. I tested a bit with themes from https://github.com/jagannatharjun/qbt-theme, and It looks identical to the screenshots.

As far as I can see, this is a straight improvement to the current situation.

Previously:

- User doesn't have Windows dark mode active: all good, qBittorrent uses default non-dark native look
- User doesn't have Windows dark mode active and they want to use custom themes:  all good, qBittorrent uses custom theme
- User has  Windows dark mode active and they want to use custom themes: all good, qBittorrent uses custom theme
- User has  Windows dark mode active and they don't want to use custom themes, they want qBittorrent to use Window's native dark theme: user is SOL

Now:

- User doesn't have Windows dark mode active: all good, qBittorrent uses default non-dark native look
- User doesn't have Windows dark mode active and they want to use custom themes:  all good, qBittorrent uses custom theme
- User has  Windows dark mode active and they want to use custom themes: all good, qBittorrent uses custom theme
- User has  Windows dark mode active and they don't want to use custom themes, they want qBittorrent to use Window's native dark theme: all good

The only "regression" is the following case, which I don't think is realistic:

- User has Windows dark mode active, don't want to use custom themes, _and_ don't want qBittorrent to use Window's native dark theme (instead, they want the default non-dark native look, which would be massively jarring).

But even in this case, there is the conf file escape hatch (these users can just set `darkmode=0` in the conf file). Other applications don't even cover this case with a conf file. It is sanely assumed that if users turn on the systemwide dark theme, they don't want a single application to be massively different from anything else. So I don't think claiming the "UX of disabling dark theme support is bad" is a relevant criticism of this change.

Final note: since with `darkmode=2` the theme switch happens without application restarts, this will work seamlessly for users who have dark mode periodically active on a schedule in Windows.

jagannatharjun commented:
personally, I don't want `darkmode=2` in the repo, this will just add another "feature" that doesn't work correctly.

FranciscoPombal commented:
@jagannatharjun 

> personally, I don't want `darkmode=2` in the repo, this will just add another "feature" that doesn't work correctly.

Disagree. Native dark theme support, even if unpolished, is better than none at all, especially when we have custom themeing support to make up for any deficiencies. As my previous comment states, now 4/4 cases are covered, without this change only 3/4. If the user really dislikes the native dark theme, they can still use custom themes, as they were forced to do previously anyway.

an0n666 commented:
Just by looking at the screenshots, it looks horrible! No one in their right mind would like this dark mode. Also if I have to turn off systemwide app dark mode for one shitty app then that's BS!

glassez commented:
Is there a way to detect current mode from within the application?

FranciscoPombal commented:
@an0n666 

> Just by looking at the screenshots, it looks horrible! No one in their right mind would like this dark mode.

Certainly less jarring for a dark theme system than the almost white normal native look. This is simpy better behavior by default.

> Also if I have to turn off systemwide app dark mode for one shitty app then that's BS!

Where did you get this idea? In my post above, I clearly detailed how all use cases are covered in a strictly better way than previously. If you don't like the native look, use a custom theme, which is what you would have done without this change in that case anyway.

> That's already on! Only qBt is failing to detect it. All other apps working fine.

Well, idk what's happening in your system. As long as those settings are toggled, and you have this new qt.conf in the same folder as the executable, it should just work. Reproduced successfully in 2 machines running 20H2 already.

@glassez 

> Is there a way to detect current mode from within the application?

I don't know, maybe. For what purpose? This change requires no additional logic.

glassez commented:
> > Is there a way to detect current mode from within the application?
> 
> I don't know, maybe. For what purpose?

IIRC, everything goes to the point that qBittorrent will have a built-in dark theme, so that we could automatically activate it when Windows dark mode is enabled.

xavier2k6 commented:
How about querying the registry? (Windows 10 >=1903)

`\HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Themes\Personalize`
`AppsUseLightTheme`
`SystemUsesLightTheme`

FranciscoPombal commented:
@glassez 

> IIRC, everything goes to the point that qBittorrent will have a built-in dark theme, so that we could automatically activate it when Windows dark mode is enabled.

We don't need to detect anything; with `darkmode=2` the change to the appropriate theme happens automatically when the Windows setting is toggled, no application restart required (https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-743331866, emphasis mine, and vice-versa for turning on):

> Dark mode won't be on by default. Support for it will be active by default. This means that if the user selects "Dark mode for applications" in the Windows settings, qBittorrent will use dark mode. **As soon as the user turns off "Dark mode for applications", qBittorrent automagically reverts back to the default native look & feel, without even needing to be restarted.**

Again, custom themes have precedence, so if the user is using a custom theme, nothing happens when the Windows setting is toggled on way or the other.

an0n666 commented:
Ok, I got it to work but still, it doesn't feel like a good option to be turned on by default. The user has no way to revert to default theme within qBt if system-wide app dark mode is on.

glassez commented:
> @glassez
> 
> > IIRC, everything goes to the point that qBittorrent will have a built-in dark theme, so that we could automatically activate it when Windows dark mode is enabled.
> 
> We don't need to detect anything; with `darkmode=2` the change to the appropriate theme happens automatically when the Windows setting is toggled, no application restart required

@FranciscoPombal 
Even though you quote my comment, you don't respond to it...

I talked about the following (assuming qBittorrent has built-in dark theme, as I said above, and no custom theme is enabled):
1. Windows Normal mode is enabled - qBittorrent uses native "look and feel".
2. Windows Dark mode is enabled - qBittorrent uses built-in dark theme.

FranciscoPombal commented:
@an0n666 

> Ok, I got it to work but still, it doesn't feel like a good option to be turned on by default. The user has no way to revert to default theme within qBt if system-wide app dark mode is on.

Yes there is, please read my above comments - you can _still_ edit `qt.conf` and set `darkmode` to `0`. But again, who exactly is going to turn on dark theme system wide in Windows and then _not_ either:
- a) use the native dark theme
- b) use a custom dark theme, if they don't like a).

?

As I explained above, this is an unrealistic use case that most (or even none) other program even provides an escape hatch for, so even by leaving the possibility of accomplishing this via editing `qt.conf` we're already doing more than enough.

@glassez

> Even though you quote my comment, you don't respond to it...
> 
> I talked about the following (assuming qBittorrent has built-in dark theme, as I said above, and no custom theme is enabled):
> 
>     1. Windows Normal mode is enabled - qBittorrent uses native "look and feel".
> 
>     2. Windows Dark mode is enabled - qBittorrent uses built-in dark theme.

If you're asking if this is what will happen, then yes, without any other changes required. If this is not what you're asking, then I guess I still don't understand your question. My comment above (https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-743341836) contains a complete break down of all cases that already answered this. I would also suggest you try this out for yourself.

---

OK, actually, here is another interpretation of your question - are you saying that you're expecting qBittorrent to have its own "native" dark theme _other than_ the styling provided by the Windows dark mode? Why would we do that? That's not "native" by definition (at least not _native to the platform_, which is what matters), and if you want something that does not fit in with the platform, there is already the custom themes mechanism for that.

99% of people just want applications to conform to the platform's theme, so as to have a consistent appearance across applications. This is what happens on Linux; if you install a dark GTK3 theme, all applications that support GTK styling (which is the vast majority, even Qt ones with the Qt style plugin package) will use it automagically, and you will have a consistent desktop experience. And this is on a platform notorious for its fragmentation, which makes this more difficult to accomplish, but it's been done.

Windows users never had this, leading to an incoherent mess of "native" themes that each application bundled (which seems to be what you want). Not as bad of an issue now as it was in the 2000s but still. macOS users seemed to suffer less from this, since the ecosystem itself has always enforced more strict conformance to a certain styling. This ended up paying off when Apple introduced macOS dark mode a while ago - very quickly most applications were able to adopt it, and it looked consistent across applications. Now, thanks to native Windows dark mode, Windows users can finally experience systemwide consistency in the appearance of their programs as well. The expectation is that the applications conform to it _by default_, not bundle a custom "native" theme!

glassez commented:
> you're expecting qBittorrent to have its own "native" dark theme _other than_ the styling provided by the Windows dark mode?

Yes.
>Why would we do that? That's not "native" by definition (at least not _native to the platform_, which is what matters), and if you want something that does not fit in with the platform, there is already the custom themes mechanism for that.

Personally, I prefer apps to look _native to the platform_, but... 
IIRC from screenshots/comments/quotes above:
1. Qt apps still don't look so _native to the platform_ in Windows Dark Mode
2. qBittorrent particularly looks ugly/unusable in Windows Dark Mode

So I suggest to provide replacement for this _native_ "look and feel" (in form of built-in dark theme that should be enabled when Windows Dark Mode is used). When Qt fix/improve _native_ dark mode "look and feel" we can/should start to use it of course.

If I did not correctly assess the background or draw the wrong conclusions, then the whole further discussion boils down to whether we should use a dark but ugly appearance, or continue using the regular but out of context appearance when Windows Dark Mode is active.

glassez commented:
And seems we're able allow user to configure default behavior in the application settings.

thalieht commented:
>99% of people just want applications to conform to the platform's theme, so as to have a consistent appearance across applications. 

Except that this just uses a "high contrast" theme which are clearly abominations in terms of aesthetics. You can get the exact same result when you activate a high contrast theme in windows settings, except this time, all apps will use the high contrast theme, not just qBt.

>This is currently experimental pending the introduction of new style that properly adapts to dark mode.

Clearly this is just a stub. Who knows whether Qt intents to fix it themselves or wait for microsoft to implement a proper dark mode.

FranciscoPombal commented:
@glassez @thalieht 

> When Qt fix/improve _native_ dark mode "look and feel" we can/should start to use it of course.

> Clearly this is just a stub. Who knows whether Qt intents to fix it themselves or wait for microsoft to implement a proper dark mode.

Although there is no timeline for Qt fixing it (as far as I'm aware), I would expect it to be sooner rather than later. Also, Microsoft _has_ implemented a "proper" dark mode (you're free to dislike their take on it of course). Seems to me implementing our own "native" theme is wasted effort.

> 2. qBittorrent particularly looks ugly/unusable in Windows Dark Mode

Still not as bad/out of place as the default non-dark native feel looks in dark mode-enabled system.

> the whole further discussion boils down to whether we should use a dark but ugly appearance, or continue using the regular but out of context appearance when Windows Dark Mode is active.

For now, yes. My argument is that _as a default_, an "ugly" preproduction dark theme is preferable to nothing even right now, when the user enables the systemwide dark mode setting.

Here is what I expect to happen with this change:
1. User turns on systemwide Windows dark mode
2a. They actually like/don't mind this preproduction native dark mode implementation, stop here.
2b. They don't like the preproduction natvive dark mode implementation, continue.
3. User installs custom theme; if step 2a did not exist, they would have done this anyway.

So, there are no extra steps. For a minority of users, there's actually less steps (the ones that like/don't mind the preproduction dark theme). And in time, the situation will solve itself as Qt improces their dark mode implementation.

FranciscoPombal commented:
Anyway, we can of course defer this until Qt fixes the dark theme. It's just that there is no downside to doing this now. Those who do not like it can use a custom theme like they would have either way. If you still think this should wait for Qt's fix, please add the `Waiting upstream` label and let's suspend this for now.

xavier2k6 commented:
**_Some more screenshots!_**

The only issues really (from my point of view) is qBittorrent's **DEFAULT Icons** aren't suitable for any type of _dark theme_ (Qt's dark mode or custom) (due to colour scheme used.......stalled arrows for example)

The enabling of Qt's dark mode makes certain things too dark but that could be rectified by a `palette` change for the whole program maybe?

## Examples
(transfer list/alternating row colour/selection bar on left pane....status/categories etc <- I'm not sure what to call it.....highlighter?)

## Windows dark mode active/CUSTOM LIGHT Theme/Qt dark mode 2 OFF
![new icons light theme Qt dark mode 2 off](https://user-images.githubusercontent.com/42386382/102026958-01895480-3d99-11eb-847f-0667a9630d1c.png)

## Windows dark mode active/CUSTOM LIGHT Theme/Qt dark mode 2 ON
![new icons light theme Qt dark mode 2](https://user-images.githubusercontent.com/42386382/102026960-03ebae80-3d99-11eb-8854-eca6b42bacf1.png)

## Windows dark mode active/CUSTOM DARK Theme/Qt dark mode 2 OFF
![new icons dark theme Qt dark mode 2 off](https://user-images.githubusercontent.com/42386382/102027005-4614f000-3d99-11eb-98c1-cdbae8e4b7e8.png)

## Windows dark mode active/CUSTOM DARK Theme/Qt dark mode 2 ON
![new icons dark theme Qt dark mode 2](https://user-images.githubusercontent.com/42386382/102027006-47deb380-3d99-11eb-8b6e-0fa7e457399e.png)

glassez commented:
> The only issues really (from my point of view) is qBittorrent's **DEFAULT Icons** aren't suitable for any type of _dark theme_

This issue has already been mentioned in some another topic. To solve it, we must provide an alternative set of icons for dark mode (just as we have provided alternative colors since time immemorial).

@FranciscoPombal 
> Again, custom themes have precedence, so if the user is using a custom theme, nothing happens when the Windows setting is toggled on way or the other.

The screenshots above prove otherwise.
It seems that if the theme does not override the palette application still follows the system settings (Dark mode). It's not a problem from my POV. We just need to understand that theme can't be considered Light/Dark if it doesn't explicitly set the palette. This also implies that the theme (which does not specify the palette explicitly) should be able to set alternative colors for the Dark mode (e.g., for transfer list), as well as alternative icons (when this is implemented in the regular appearance).

xavier2k6 commented:
>This issue has already been mentioned in some another topic. To solve it, we must provide an alternative set of icons for dark mode (just as we have provided alternative colors since time immemorial).

I already mentioned the other topic/PR, you may have overlooked it.

It's PR https://github.com/qbittorrent/qBittorrent/pull/12965

FranciscoPombal commented:
@glassez 

> The screenshots above prove otherwise.
It seems that if the theme does not override the palette application still follows the system settings (Dark mode). It's not a problem from my POV. We just need to understand that theme can't be considered Light/Dark if it doesn't explicitly set the palette. This also implies that the theme (which does not specify the palette explicitly) should be able to set alternative colors for the Dark mode (e.g., for transfer list), as well as alternative icons (when this is implemented in the regular appearance).

The screenshots in https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-744088206 don't disprove my assertions completely, they merely reveal that there might be a special case where they might not hold (that of when a theme does not override the pallete). I experienced the case of screenshots 1,3 and 4 in my testing (which show that the mechanism of "theme selection" works as I have previously described). I did not observe the case of screenshot 2. Perhaps precisely because I did not test with a theme that did not override the palette.

I agree that we shouldn't consider this a problem. But we should either: a) document this and encourage custom theme developers to specify a custom palette, explaining that if they don't do this, the dark/light palette of the underlying platform theme will be used or b) mandate that a palette always be specified (either a custom or explicitly deferring to the underlying one) to completely prevent this.

About the icons: we don't need an alternative set necessarily. It is possible to have a single set - there may be color combinations that work just fine on both light and dark theme (see the solarized themes for an example of this, the chosen foreground colors look good on both backgrounds - https://ethanschoonover.com/solarized/, https://en.wikipedia.org/wiki/Solarized_(color_scheme)). But it takes someone with actual design skills to figure this out. As mentioned above, I believe https://github.com/qbittorrent/qBittorrent/pull/12965 was attempting to accomplish this.

glassez commented:
> b) mandate that a palette always be specified (either a custom or explicitly deferring to the underlying one) to completely prevent this.

👎 
Generally I don't see any good reason to prohibit the theme from following the system palette.


> About the icons: we don't need an alternative set necessarily. It is possible to have a single set - there may be color combinations that work just fine on both light and dark theme.
But it takes someone with actual design skills to figure this out. 

This is why it is easier to provide the ability to have different sets for light/dark mode. It should not prevent you from using only one universal set, if someone has enough talent to create one.

FranciscoPombal commented:
@glassez 

>👎 
>Generally I don't see any good reason to prohibit the theme from following the system palette.

Perhaps I did not express myself clearly enough; in that option I'm not saying we should outright "prohibit the theme from following the system palette", I'm saying that we should prohibit it from doing so _implicitly_. The theme may use the system palette or a custom one, but it must be explicit about that. Not sure if and how that can be enforced, but it's just a suggestion.

nowshed-imran commented:
My experience says having two icon set will be the best case scenario. Every Google app toggles its icon color when going to the dark mode. Pleasing both theme with one icon set will just not work. All application icons combined sized at around 650 KB! 
I just don't understand what is the barrier. Here is my proposal,
At this location `qBittorrent/src/icons/ `, there will be three folders,
1. Flags,
2. Dark,
3. Light
Based on theme icons will automatically selected. User can change it from the settings too. This is a modern solution.
If someone can help me with the code and seriously reviews my PR so that it might merge, I am willing to work on it again.
Decision is yours to take.

FranciscoPombal commented:
@glassez @Chocobo1 @sledgehammer999 

I think we all agree that Qt's current built-in dark theme for Windows is quite crude, and only suitable for fans of high contrast, but it is better than nothing. Some users would actually like it - consider that recently even GitHub has introduced a new high-contrast dark theme due to popular demand.

I think it's a waste for this PR to be sitting here all this time, instead of empowering users to choose. It doesn't have to be perfect, it's completely fine to state in the changelog that native dark theme support on Windows is experimental and the sole responsibility of Qt itself.

xavier2k6 commented:
> Is there a way to detect current mode from within the application?

Something like below may help:
https://github.com/oleksis/picta-dl-gui/discussions/3#discussioncomment-219576

switching to `fusion style` by default would also fix a few issues that have been reported.

We could have 2 `Palette's` light/dark

xavier2k6 commented:
>I think we all agree that Qt's current built-in dark theme for Windows is quite crude

This may have been changed in newer revisions of Qt6.0.x/Qt6.1.x (Haven't tested)

glassez commented:
> I think it's a waste for this PR to be sitting here all this time, instead of empowering users to choose. It doesn't have to be perfect, it's completely fine to state in the changelog that native dark theme support on Windows is experimental and the sole responsibility of Qt itself.

Well, we could provide it to users to receive real feedbacks. But we should definitely revert it back if feedbacks are mostly negative.

jagannatharjun commented:
-1 from me, main problem being

>User has Windows dark mode active, don't want to use custom themes, and don't want qBittorrent to use Window's native dark theme

Most of the people have dark theme enabled (Atleast most of the people I know), This options will show horrendous window that I no one in their right mind would ever use and will just scare away people, IMO the default should be acceptable, not something experimental for app with more than 100 million downloads. Also - 

>if I have to turn off systemwide app dark mode for one shitty app then that's BS!

FranciscoPombal commented:
@jagannatharjun 

> -1 from me, main problem being
> 
> > User has Windows dark mode active, don't want to use custom themes, and don't want qBittorrent to use Window's native dark theme

I presented this example specifically as an example of an unrealistic scenario, which has an easy fix any way: comment out the setting in qBittorrent's config file.
 
> Most of the people have dark theme enabled (Atleast most of the people I know), This options will show horrendous window that no one in their right mind would ever use and will just scare away people, IMO the default should be acceptable, not something experimental for app with more than 100 million downloads. Also -

People use dark themes to save their eyes. To provide my own bit of anecdotal evidence, everyone I know who uses dark themes prefers even a shitty one to none at all. And with the ubiquity of AMOLED phones, high contrast dark themes have found their way into the mainstream as of late.

> if I have to turn off systemwide app dark mode for one shitty app then that's BS!

FUD. Again, you can comment out the setting in the config file...

glassez commented:
> -1 from me, main problem being
> 
> > User has Windows dark mode active, don't want to use custom themes, and don't want qBittorrent to use Window's native dark theme
> 
> Most of the people have dark theme enabled (Atleast most of the people I know), This options will show horrendous window that no one in their right mind would ever use and will just scare away people, IMO the default should be acceptable, not something experimental for app with more than 100 million downloads. Also -
> 
> > if I have to turn off systemwide app dark mode for one shitty app then that's BS!

Well, it looks like the negative feedbacks are here even before it could be merged...

glassez commented:
Maybe we could still provide it optionally through some advanced setting? Or just add it commented out and mention somewhere that someone can uncomment it if they want to experiment with it?

xavier2k6 commented:
@FranciscoPombal @jagannatharjun Maybe we can look into a proper native dark theme as I pointed out in https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-873563009

We need to change the styling on windows anyway.

jagannatharjun commented:
> I presented this example specifically as an example of an unrealistic scenario, 

This is not an unrealistic scenario, I repeat _no one in their right mind would ever use this_

> which has an easy fix anyway: comment out the setting in qBittorrent's config file.

Any answer like ` comment out the setting in qBittorrent's config file` for a default behaviour is not a good answer.

> FUD. Again, you can comment out the setting in the config file...

if you want this themeing you can run qbittorrent like `qbittorrent -platform windows:darkmode=2`, this would make your highly technical ideal user very happy.

FranciscoPombal commented:
@glassez 

> Well, it looks like the negative feedbacks are here even before it could be merged...

FUD would be a more accurate term. I have already addressed the raised concerns.

> Maybe we could still provide it optionally through some advanced setting? Or just add it commented out and mention somewhere that someone can uncomment it if they want to experiment with it?

I still think `(number of users who would be angry at the dark theme because they think it's ugly and would want to turn it off) << (number of users who would never discover the setting if it were commented out by default, even though they would have liked the theme)`.

@jagannatharjun 

> This is not an unrealistic scenario, I repeat _no one in their right mind would ever use this_

Citation needed. Look, we both provided nothing more than anecdotal evidence, but I'm still seriously willing to bet that the majority of people will take an "ugly" dark theme over getting their eyes blasted with the light of 1000 suns at night, _by default_.

> Any answer like ` comment out the setting in qBittorrent's config file` for a default behaviour is not a good answer.

If your aesthetically-minded friends are so concerned about it, they can install an custom theme anyway. It's acceptable to be turned on by default in this case, because, again, I'm sure logical people who find the native dark theme ugly would first try out a custom dark theme rather than trying to revert to using the light theme, given that the rest of their system is using a dark theme.

> if you want this themeing you can run qbittorrent like `qbittorrent -platform windows:darkmode=2`, this would make your highly technical ideal user very happy. They should be able to do this quite easily. You won't even need to change anything in this repo.

You've got it backwards, I want to make the not-so technically minded users happy, by having at least a primitive dark theme "just work". Those with more peculiar needs can use custom themes or even edit the config file. Layman users will never run qbittorrent that way (`qbittorrent -platform windows:darkmode=2`).

FranciscoPombal commented:
@xavier2k6 

> @FranciscoPombal @jagannatharjun Maybe we can look into a proper native dark theme as I pointed out in [#13952 (comment)](https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-873563009)
> 
> We need to change the styling on windows anyway.
> 
> I'm sure we can make the registry check behind a clickable button or something so it can be toggled on/off etc.

This looks promising. Could be the next iteration - but until it materializes, users could have already have had something to play with if this had been merged... instead we've bickered for 6 months because the theme is not perfect lol. This is an iterative process, people!

glassez commented:
>Look, we both provided nothing more than anecdotal evidence

@FranciscoPombal 
Do you use Windows dark theme daily?
@jagannatharjun, do you?

jagannatharjun commented:
> @FranciscoPombal
Do you use Windows dark theme daily?

Francis don't even use windows. 

> @jagannatharjun, do you?

I use UI theme bundle.

FranciscoPombal commented:
@glassez 

> @FranciscoPombal
> Do you use Windows dark theme daily?

Yes, both in a machine where I have Windows installed on bare-metal, and on a VM that I have on one of my Linux systems, which is also running a dark theme (xfce + `adwaita-dark`). I have also been running AMOLED-dark themes on my phones as soon as that became an option.

FranciscoPombal commented:
@jagannatharjun 

> Francis don't even use windows.

Unfortunately, this is not true - I am a sinner because of both gaming and the occasional cross-platform stint (as is the case for qBittorrent).

glassez commented:
> > @FranciscoPombal
> > Do you use Windows dark theme daily?
> 
> Francis don't even use windows.
> 
> > @jagannatharjun, do you?
> 
> I use UI theme bundle.

Then it's like an argument between a mute and a deaf person, sorry...
For me personally, it deserves nothing more than to be mentioned in the Wiki. But I can only approve it in order to collect some feedback statistics from real users.

FranciscoPombal commented:
@glassez 

> For me personally, it deserves nothing more than to be mentioned in the Wiki. But I can only approve it in order to collect some feedback statistics from real users.

Merging this would go a long way to get users to provide feedback... If you are concerned about users complaining, the responsibility is not solely yours, you can tell them (and they can see) that it was originally my idea. But perhaps I'm more optimistic in thinking that users would be most likely to provide constructive feedback, rather than come at us with pitchforks in case they don't like it. Anyway, you know I'll do my best to be present and active in addressing feedback from the theme. This is Free Software made by the people, for the people, one either engages productively in the discussion or gets their post closed & locked and we all move on.

FranciscoPombal commented:
@glassez 

> Then it's like an argument between a mute and a deaf person, sorry...

Not sure what he meant by "UI theme bundle", but contrary to popular belief, I do sometimes find myself using Windows, and when I do so, I always use the native dark theme FWIW: https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-873643189, https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-873643661

glassez commented:
In any case, my approval and its reasons are set out above. But this is not enough, as you know. And I have no reason to put pressure on someone else to get this PR approved.

jagannatharjun commented:
> Then it's like an argument between a mute and a deaf person, sorry...

My main OS is always been windows and I only started using linux recently because of work.


> I can only approve it in order to collect some feedback statistics from real users.

This PR would've been fine if it were to be not make this the default behavior. My concern is providing an experimental feature as a default behavior with no clear way to change it. Plus it doesn't look good.

>  it deserves nothing more than to be mentioned in the Wiki.

👍 


@FranciscoPombal 

> Not sure what he meant by "UI theme bundle"

I meant by qbittorrent's UI theme system.

sakkamade commented:
Sorry, I don't see the point in this argument...

If someone wishes a dark theme, presently their only choice is to try the custom theme.
If someone uses dark theme in Windows, they will undoubtedly try the custom dark theme—for what is the point in having only qbit blindly white, right?

Whereas after this PR, if someone uses dark theme in Windows, they would be furnished with power to choose:
1. Use ugly default dark theme
    - And if someone hates it from the bottom of the heart, it may even be disabled via config file.
2. Try out a custom dark theme

Then why be against it at all?

---

> > Not sure what he meant by "UI theme bundle"
> 
> I meant by qbittorrent's UI theme system.

I guess it means custom theme...?

sakkamade commented:
> Then why be against it at all?

Simultaneously, I can understand that there is no point in feature which no one intends to make use of.

sakkamade commented:
Someone may say "You don't use dark them. how you could speak of it?" however I need not be a dark-theme specialist to understand that to improve something afterwards is much easier than to create something flawless right from the start...

sakkamade commented:
from https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-873638953
> if you want this themeing you can run qbittorrent like `qbittorrent -platform windows:darkmode=2`

@jagannatharjun,
 However easy it may seem to you, here is some example of workaround which people suggests https://github.com/qbittorrent/qBittorrent/issues/11917#issuecomment-817115258.

xavier2k6 commented:
If this was to be merged, I presume it would only be in `master` for `4.4.x beta builds` & not backported to `4.3.x branch`for upcoming release?

FranciscoPombal commented:
@xavier2k6 

> If this was to be merged, I presume it would only be in `master` for `4.4.x beta builds` & not backported to `4.3.x branch`for upcoming release?

No reason why it couldn't be backported.

sakkamade commented:
It seems my commentaries left no as deep impression here as I wanted them to, so I shall make myself more plain and elaborate on [this point](https://github.com/qbittorrent/qBittorrent/pull/13952#issuecomment-873673853).

Whilst Windows user has the custom themes as the only native option of dark theming, they will not even dare to dwell upon the thought that there might be another direction, and even more native, too.
And whereas no common knowledge therein, sparsely there would ever be any progress to it, as the contributors would also be gravitated upon improving but custom themes, for this have become a default route by now.

xavier2k6 commented:
If we were to make this work to some degree.....we would have to fix icon colors etc. that would work for both normal & dark mode.

See a WIP example:

![my icons light](https://user-images.githubusercontent.com/42386382/132944796-441a25bd-352d-4023-b88e-3ceff899b5f1.png)
![my icons dark](https://user-images.githubusercontent.com/42386382/132944797-d46dc208-8354-4ca6-9ce4-e4b1efb034a8.png)

ArcticGems commented:
> 
> 
> If we were to make this work to some degree.....we would have to fix icon colors etc. that would work for both normal & dark mode.
> 
> See a WIP example:

Having 2 default themes would be great (+ a toggle option) and it would make it easier for @nowshed-imran since he could make 2 sets, one for light/normal and one for dark.

xavier2k6 commented:
>Having 2 default themes would be great (+ a toggle option) and it would make it easier for @nowshed-imran since he could make 2 sets, one for light/normal and one for dark.

There's no need for two sets of icons.......if we get the coloring right with `icons` & `text`

The toggle option you refer to is what this PR does.......activates windows dark mode for qBittorrent.

ArcticGems commented:
> 
> There's no need for two sets of icons.......if we get the coloring right with `icons` & `text`

Wasn't that one of the reasons why people couldn't agree, because it's hard to make it work for both light & dark themes?

xavier2k6 commented:
>Wasn't that one of the reasons why people couldn't agree, because it's hard to make it work for both light & dark themes?

Look at the screen shots from my previous post........the icon colors are the same.

I use a different source to check out colors which give an example of what it looks like with white background/black background.

Example:
![Forest Green](https://user-images.githubusercontent.com/42386382/131481207-ca556c48-d0fb-448b-94c0-6d1fce5f3149.png)

debiedowner commented:
Wow... For the last two years, my tedious routine has been to enable [a custom dark theme](https://gitlab.com/qbittorrent-black-theme/client) every evening and restart qBittorrent, and disable it in the morning and then again restart. Except sometimes in the evening I didn't want to stop a download by restarting, so I had to suffer the blinding white UI until I could restart. I never knew that all that effort and discomfort was unnecessary, that qBittorrent can actually change automatically according to Windows theme, without even needing to restart... And some qBittorrent contributors were blocking this feature, because they personally disliked the aesthetics of this dark theme... Incredible.

Anyway, I was happy to see that this change doesn't require recompiling and I can simply edit the `qt.conf` file on my system. Which I did immediately, and it looks and works great. The only problem is that when transitioning from dark to light theme, some text in qBittorrent options dialog remain white and difficult to read, which requires a restart to correct. But it's no big deal as I don't look at settings that often, and even if I did actually need to restart to see them, it is still a very significant improvement over the existing situation where I need to restart every time for theme change to take effect. Besides, restarting in the morning is less of a problem for me; it was having to restart in the evening (i.e. transitioning from light to dark theme) that was annoying, since I am more likely to be in the middle of something.

Also, interestingly, this dark theme is slightly more compact than qBittorrent's default theme, but only if qBittorrent was started while Windows is already set to dark mode, not transitioned while it is open. I figured this is due to the changes in version 4.2.2, so I tried [4.2.1's compact theme](https://github.com/FoxInFlame/legacy-qBittorrent) to see if that's the case, but 4.2.1 theme is even more compact; dark theme is somewhere in between. Anyway, it's a minor difference, I only noticed it because I took screenshots to compare it with the default theme and the custom dark theme I used to use until today.

Also, I am using Qt 5.15.2; I didn't test with Qt 6 version of qBittorrent 4.4.0, maybe that makes improvements to the aesthetics aspect that might satisfy qBittorrent contributors. I didn't check as I think this version, as seen in the screenshots above, is already perfectly fine, much better than all the [unofficial themes listed in the wiki](https://github.com/qbittorrent/qBittorrent/wiki/List-of-known-qBittorrent-themes) (except for the "legacy-qBittorrent" and "qbittorrent-black-theme"), in addition to the ease-of-use advantages.
